src/openms/include/OpenMS/METADATA/ID/ObservationMatch.h

```diff
@@ -0,0 +1,154 @@
+// ETH Zurich, and Freie Universitaet Berlin 2002-2020.
+//
+// This software is released under a three-clause BSD license:
+//  * Redistributions of source code must retain the above copyright
+//    notice, this list of conditions and the following disclaimer.
+//  * Redistributions in binary form must reproduce the above copyright
+//    notice, this list of conditions and the following disclaimer in the
+//    documentation and/or other materials provided with the distribution.
+//  * Neither the name of any author or any participating institution
+//    may be used to endorse or promote products derived from this software
+//    without specific prior written permission.
+// For a full list of authors, refer to the file AUTHORS.
+// --------------------------------------------------------------------------
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL ANY OF THE AUTHORS OR THE CONTRIBUTING
+// INSTITUTIONS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+// OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+// WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
+// OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
+// ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// --------------------------------------------------------------------------
+// $Maintainer: Hendrik Weisser $
+// $Authors: Hendrik Weisser $
+// --------------------------------------------------------------------------
+
+#pragma once
+
+#include <OpenMS/METADATA/ID/Observation.h>
+#include <OpenMS/METADATA/ID/MetaData.h>
+#include <OpenMS/METADATA/ID/IdentifiedMolecule.h>
+#include <OpenMS/METADATA/PeptideHit.h> // for "PeakAnnotation"
+#include <OpenMS/CHEMISTRY/AdductInfo.h>
+
+#include <boost/multi_index_container.hpp>
+#include <boost/multi_index/ordered_index.hpp>
+#include <boost/multi_index/composite_key.hpp>
+
+namespace OpenMS
+{
+  namespace IdentificationDataInternal
+  {
+    // @TODO: move "PeakAnnotation" out of "PeptideHit"
+    typedef std::vector<PeptideHit::PeakAnnotation> PeakAnnotations;
+    typedef std::map<boost::optional<ProcessingStepRef>,
+                     PeakAnnotations> PeakAnnotationSteps;
+
+    /// Comparator for adducts
+    // @TODO: this allows adducts with duplicate names, but requires different
+    // sum formulas/charges - is this what we want?
+    struct AdductCompare
+    {
+      bool operator()(const AdductInfo& left, const AdductInfo& right) const
+      {
+        return (std::make_pair(left.getCharge(), left.getEmpiricalFormula()) <
```

Comment:
this is a bit inefficient?!
use `std::tie`?

Reply:
can you tie to a return type? not sure

Reply:
I'm pretty sure you can use `tie` with the comparison operator, but is it more efficient than `make_pair`?

Reply:
Tie with the comparison operator is not an issue but the parameters passed to tie. I think these might not be temporaries. Will look it up…